### ISSUE TITLE
`KeyError` raised in `format_skeleton`  when using fuzzy matching

### ISSUE THREAD

**Issue by tomasr8** [python-babel/babel#1084]
## Overview Description

When using `format_skeleton`, a `KeyError` is raised despite using `fuzzy=True`.

## Steps to Reproduce

```python3
from datetime import datetime

from babel.dates import format_skeleton


dt = datetime(2012, 1, 1, 14, 30, 59)
format_skeleton("G", dt, locale="cs_CZ", fuzzy=True)
```

From the docs:

> fuzzy – If the skeleton is not found, allow choosing a skeleton that’s close enough to it.

The way I read it is that as long as I pass `fuzzy=True` (which is the default) a skeleton should _always_ be found and I should not need to worry about `KeyError`s in that case.

Even the example in the docs makes it seem a `KeyError` is only thrown with `fuzzy=False`:
```python3
format_skeleton('yMMd', t, fuzzy=False, locale='fi')  # yMMd is not in the Finnish locale, an error is thrown
```

If this is the expected behaviour, I think the docs should state that explicitly. If not, I would suggest adding `allow_different_fields=True` to the underlying `match_skeleton()` call.

## Additional Information
Babel version: `2.15`

**Comment by tomasr8** [comment]
@akx What do you think? Should we fix this?

**Comment by akx** [comment]
Yeah, I think so. My original commit for fuzzy skeletons (fuzzy skeletons, yikes what a horrible mental image) https://github.com/python-babel/babel/commit/cd70395b0f138b7f60b0866f5b2b164f1a025786 refers to [the ICU4J `getBestSkeleton` function](https://github.com/unicode-org/icu/blob/ee93218eabd72406d81f8a1c92a6b9249ceb4de9/icu4j/main/core/src/main/java/com/ibm/icu/text/DateIntervalInfo.java#L1081) as the reference implementation.

Apparently there's been at least one commit https://github.com/unicode-org/icu/commit/edaebfa64ee673ed580f464123d3afd4deffc693 that has improved that function since – we should maybe look at whether that improvement would in fact fix this?

**Comment by tomasr8** [comment]
Looks like the only thing that changed is more format specifier replacements which we could add as well:

```python
    if 'k' in skeleton and not any('k' in option for option in options):
        skeleton = skeleton.replace('k', 'H')
    if 'K' in skeleton and not any('K' in option for option in options):
        skeleton = skeleton.replace('K', 'h')
    if 'a' in skeleton and not any('a' in option for option in options):
        skeleton = skeleton.replace('a', '')
    if 'b' in skeleton and not any('b' in option for option in options):
        skeleton = skeleton.replace('b', '')
```

This doesn't fix the `KeyError` issue though. I'm a bit unsure whether we _should_ fix it or just make it clear in the docs that the function can in fact raise even when `fuzzy=True`. If we do want to fix this, I think we can implement the recommendation from the tr35 doc for missing fields: https://www.unicode.org/reports/tr35/tr35-72/tr35-dates.html#missing-skeleton-fields

**Comment by akx** [comment]
For the time being, we could maybe just document that `fuzzy=True` may still raise, and fix that in a subsequent version using that recommendation?

**Comment by geonseok-ship-it** [comment]
Is this still open for a contribution? I'd like to take it.

Reading the thread, you landed on documenting that `fuzzy=True` can still raise, and doing the real fix later. @tomasr8's comment already has the ICU4J specifier replacements pasted, so both halves look tractable — should I do them as one PR or keep the doc note separate so it can ship immediately?

One thing I'd check before writing anything: whether adding the `k`/`K`/`a` replacements is enough to make `format_skeleton("G", ..., fuzzy=True)` stop raising, or whether `G` needs handling of its own since it has no time-related fallback. Will confirm against the actual data before proposing anything.

**Comment by geonseok-ship-it** [comment]
Following up on my own offer — both halves you agreed on are already on main, so there's nothing left to implement from that plan:

- the `fuzzy=True` docs note landed in #1106 (2024-07-25)
- @tomasr8's extra specifier replacements landed in #1122 (2024-10-19)

But the reported case still raises. I checked against current main and `format_skeleton("G", dt, locale="cs_CZ", fuzzy=True)` still gives `KeyError: None`.

The reason the #1122 recipe doesn't reach it: `"G"` contains none of `z`, `k`, `K`, `a` or `b`, so every replacement is a no-op for this skeleton. The failure is one level up — `match_skeleton` discards candidates whose field set differs, and `"G"` shares no field with any available option. ICU4J doesn't hit this because when `getBestSkeleton` finds nothing it falls back to `DateTimePatternGenerator`, which synthesises a pattern from the locale's available formats (tr35). Babel has no equivalent.

So closing the gap properly means either implementing that fallback, or deciding that `fuzzy` is best-effort and a skeleton sharing no fields with any option legitimately raises — which is arguably what the #1106 docs note already says.

Happy to take either direction if you'd like it, or to close this if the docs note is considered the resolution.
